File: Ocaml/abb.ml
(* Note: this file results automatically in a module implementation
        named Abb *)

(*------Auxiliar functions to work with pointers in Ocaml-------*)

type 'a pointer = Null | Pointer of 'a ref;;

let ( !^ ) = function
    | Null -> invalid_arg "Attempt to dereference the null pointer"
    | Pointer r -> !r;;

let ( ^:= ) p v =
    match p with
    | Null -> invalid_arg "Attempt to assign the null pointer"
    | Pointer r -> r := v;;

let new_pointer x = Pointer (ref x);;

(****************************************************************)

type tKey = int;;
type tPos = tNode pointer
and tNode = Empty | Node of nodeStruct
and nodeStruct = {
    mutable key: tKey; 
    mutable left: tPos;
    mutable right: tPos};;
type tBST = tPos;;

(****************************************************************)

(* emptyTree does not have the same usage as the pascal version
    because t has to be declared before calling the function,
    so its only use is changing the value of t to Empty *)
let emptyTree (t: tBST) = 
    t ^:= Empty;;

(* createNode would be useless since you need to initialize
    t before calling the function, and you can't do something
    like "let createNode = new_pointer Empty" because this will
    create only one pointer called "createNode" the moment you 
    define the function. *)

let root (t: tBST) = match (!^t) with
    Node {key = k; left = l; right = r} -> k;;

let leftChild (t: tBST) = match (!^t) with
    Node {key = k; left = l; right = r} -> l;;

let rightChild (t: tBST) = match (!^t) with
    Node {key = k; left = l; right = r} -> r;;

let isEmptyTree (t: tBST) =
    (!^t == Empty);;

(****************************************************************)

let rec insert_r (t: tBST) x = match (!^t) with
    Empty -> 
        t ^:= Node {
            key = x;
            left = new_pointer Empty;
            right = new_pointer Empty}
|   Node {key = k; left = l; right = r} ->
        if (x < k) then
            insert_r l x
        else if (x > k) then
            insert_r r x;;

let insert_i (t: tBST) x =
    let newNode = new_pointer (Node {
        key = x;
        left = new_pointer Empty;
        right = new_pointer Empty
    }) in
    if (isEmptyTree t) then
        t ^:= !^newNode
    else
        let father = new_pointer (new_pointer Empty) in
        let child = new_pointer t in
        while ((not (isEmptyTree (!^child))) && (root (!^child) != x)) do
            father ^:= !^child;
            if (x < root (!^child)) then
                child ^:= leftChild (!^child)
            else
                child ^:= rightChild (!^child)
        done;
        if (isEmptyTree (!^child)) then
            if (x < (root (!^father))) then
                !^father ^:= Node {
                    key = root (!^father);
                    left = newNode;
                    right = rightChild (!^father)
                }
            else
                !^father ^:= Node {
                    key = root !^(father);
                    left = leftChild (!^father);
                    right = newNode
                }
        else ();;

let insertKey (t: tBST) x =
    insert_r t x;;

(****************************************************************)

let rec search_r (t: tBST) x =
    if (isEmptyTree t) then
        new_pointer Empty
    else if ((root t) == x) then
        t
    else if (x < (root t)) then
        search_r (leftChild t) x
    else
        search_r (rightChild t) x;;

let search_i (t: tBST) x =
    let node = new_pointer t in
    while ((not (isEmptyTree (!^node))) && (root (!^node) != x)) do
        if (x < root (!^node)) then
            node ^:= leftChild (!^node)
        else
            node ^:= rightChild (!^node)
    done;
    !^node;;

let searchKey (t: tBST) x =
    search_r t x;;

(****************************************************************)

let rec remove_r (t: tBST) x = match (!^t) with
    Empty -> ()
|   Node {key = k; left = l; right = r} ->
        if (x < k) then
            remove_r l x
        else if (x > k) then
            remove_r r x
        else 
            let (aux: tBST) = t in
        if (isEmptyTree l)
            then t ^:= !^r
        else if (isEmptyTree r)
            then t ^:= !^l
        else
    let rec sup2 (t2: tBST) = 
        if (not (isEmptyTree (rightChild t2))) then
            sup2 (rightChild t2)
        else
            (aux ^:= Node {
                key = root t2;
                left = leftChild aux;
                right = rightChild aux
            };
            t2 ^:= !^(leftChild t2))
    in
        sup2 l;;

let remove_i (t: tBST) x =
    let childNum = new_pointer 0 in
    let sup = new_pointer t in
    let fSup = new_pointer (new_pointer Empty) in
    (* search for the node to remove *)
    while ((not (isEmptyTree (!^sup))) && (root (!^sup) != x)) do
        fSup ^:= !^sup;
        if (x < root (!^sup)) then
                sup ^:= leftChild (!^sup)
            else
                sup ^:= rightChild (!^sup)
    done;
    (* if it isn't in t, do nothing *)
    if (not (isEmptyTree (!^sup))) then (
        (* count the number of childs *)
        if (not (isEmptyTree (leftChild (!^sup)))) then
            childNum ^:= (!^childNum + 1)
        else ();
        if (not (isEmptyTree (rightChild (!^sup)))) then
            childNum ^:= (!^childNum + 1)
        else ();
        match (!^childNum) with
            0 -> (* remove the leaf node *)

                if (isEmptyTree(!^fSup)) then
                    t ^:= Empty (* the root was the only node of the tree *)
                else if (leftChild (!^fSup) == !^sup) then
                    leftChild (!^fSup) ^:= Empty
                else 
                    rightChild (!^fSup) ^:= Empty
        |   1 -> (* remove node with only 1 child *)
                let childNotEmpty = new_pointer Empty in
                if (isEmptyTree (leftChild (!^sup))) then 
                    childNotEmpty ^:= !^(rightChild (!^sup))
                else
                    childNotEmpty ^:= !^(leftChild (!^sup));
                if (isEmptyTree(!^fSup)) then
                    (* we remove the root *)
                    t ^:= !^childNotEmpty
                else if (leftChild (!^fSup) == (!^sup)) then
                    leftChild (!^fSup) ^:= !^childNotEmpty
                else 
                    rightChild (!^fSup) ^:= !^childNotEmpty

        |   _ -> (* remove node with 2 childs *)
                fSup ^:= !^sup;
                let supLeftMax = new_pointer (leftChild (!^sup)) in
                while (not (isEmptyTree(rightChild (!^supLeftMax)))) do
                    fSup ^:= !^supLeftMax;
                    supLeftMax ^:= rightChild (!^supLeftMax)
                done;
                !^sup ^:= Node {
                    key = root !^(supLeftMax);
                    left = leftChild (!^sup);
                    right = rightChild (!^sup)
                };
                if ((!^fSup) == (!^sup)) then
                    leftChild (!^fSup) ^:= !^(leftChild (!^supLeftMax))
                else 
                    rightChild (!^fSup) ^:= !^(leftChild (!^supLeftMax))
    )
    else ();;

let removeKey (t: tBST) x =
    remove_r t x;;

File: Python/abb.py
# python considers this file a module automatically
class Pos:
	def __init__(self, obj):
		self.obj = obj
	def get(self):
		return self.obj
	def set(self, obj):
		self.obj = obj

class Node:
	def __init__(self, key):
		self.key = key
		self.leftChild = Pos(None)
		self.rightChild = Pos(None)


class BinarySearchTree:
	def __init__(self, root = Pos(None)):
		self.root = root

#----------------------------------------------------------------

def insertKey(bst, key):
	if (bst.root.get() is None):
		bst.root.set(Node(key))
	else:
		insert_i(bst.root, key)

def insert_r(currentNode, key):
	if key < currentNode.get().key:
		if not (isEmpty(currentNode.get().leftChild.get())):
			insert_r(currentNode.get().leftChild, key)
		else:
			currentNode.get().leftChild.set(Node(key))
	elif key > currentNode.get().key:
		if not (isEmpty(currentNode.get().rightChild.get())):
			insert_r(currentNode.get().rightChild, key)
		else:
			currentNode.get().rightChild.set(Node(key))
	# ignore duplicates

def insert_i(currentNode, key):
	father = Pos(None)
	child = currentNode
	while (child.get() is not None) and (child.get().key != key):
		father = child
		if key < child.get().key:
			child = child.get().leftChild
		else:
			child = child.get().rightChild
	if (child.get() is None):
		if (key < father.get().key):
			father.get().leftChild.set(Node(key))
		else:
			father.get().rightChild.set(Node(key))
	# ignore duplicates

#----------------------------------------------------------------

def searchKey(bst, key):
	return search_i(bst.root, key)

def search_r(currentNode, key):
	if (isEmpty(currentNode.get())):
		return BinarySearchTree(Pos(None))
	elif currentNode.get().key == key:
		return BinarySearchTree(currentNode)
	elif key < currentNode.get().key:
		return search_r(currentNode.get().leftChild, key)
	else:
		return search_r(currentNode.get().rightChild, key)

def search_i(currentNode, key):
	node = currentNode
	while (not isEmpty(node.get()) and (node.get().key != key)):
		if key < node.get().key:
			node = node.get().leftChild
		else:
			node = node.get().rightChild
	return BinarySearchTree(node)

#----------------------------------------------------------------

def leftChild(node):
	return node.get().leftChild

#----------------------------------------------------------------

def rightChild(node):
	return node.get().rightChild

#----------------------------------------------------------------

def root(node):
	return node.root.get().key

#----------------------------------------------------------------

def isEmpty(bst):
	return bst is None

#----------------------------------------------------------------

def remove_r(currentNode, key):
	aux = Pos(None)
	# procedure to be called when the node to be removed
	# has two childs, to replace it with the greatest key
	# from its left subtree.
	def sup2(bst):
		if (bst.get().rightChild.get() is not None):
			sup2(bst.get().rightChild)
		else:
			aux.get().key = bst.get().key
			bst.set(bst.get().leftChild.get())
	# if its not in the tree, nothing is done
	if (currentNode.get() is not None):
		if (key < currentNode.get().key):
			remove_r(currentNode.get().leftChild, key)
		elif (key > currentNode.get().key):
			remove_r(currentNode.get().rightChild, key)
		else:
			aux.set(currentNode.get())
			if currentNode.get().leftChild.get() is None:
				currentNode.set(currentNode.get().rightChild.get())
			elif currentNode.get().rightChild.get() is None:
				currentNode.set(currentNode.get().leftChild.get())
			else:
				sup2(currentNode.get().leftChild)

def remove_i(currentNode, key):
	sup = currentNode	# node to be removed
	fSup = Pos(None)	# father of the node to be removed
	# search for the node to be removed
	while (not isEmpty(sup.get())) and (sup.get().key != key):
		fSup = sup
		if (key < sup.get().key):
			sup = sup.get().leftChild
		else:
			sup = sup.get().rightChild
	# if its not in the tree, nothing is done
	if (not isEmpty(sup.get())):
		# count the number of childs
		childNum = 0
		if (not isEmpty(sup.get().leftChild.get())):
			childNum += 1
		if (not isEmpty(sup.get().rightChild.get())):
			childNum += 1
		# remove leaf node
		if (childNum == 0):
			if (isEmpty(fSup.get())):
				# the root was the only node of the tree
				currentNode.set(None)
			elif (fSup.get().leftChild.get() == sup.get()):
				fSup.get().leftChild.set(None)
			else:
				fSup.get().rightChild.set(None)
		# remove node with only one child
		elif (childNum == 1):
			if (isEmpty(sup.get().leftChild.get())):
				childNotEmpty = sup.get().rightChild
			else:
				childNotEmpty = sup.get().leftChild
			if (isEmpty(fSup.get())):
				# the root is removed
				currentNode.set(childNotEmpty.get())
			elif (fSup.get().leftChild.get() == sup.get()):
				fSup.get().leftChild.set(childNotEmpty.get())
			else:
				fSup.get().rightChild.set(childNotEmpty.get())
		# remove node with two childs
		else:
			fSup = sup
			# supLeftMax: node with the greatest value from the
			# left subtree
			supLeftMax = sup.get().leftChild
			while (not isEmpty(supLeftMax.get().rightChild.get())):
				fSup = supLeftMax
				supLeftMax = supLeftMax.get().rightChild
			sup.get().key = supLeftMax.get().key
			if (fSup.get() == sup.get()):
				fSup.get().leftChild.set(supLeftMax.get().leftChild.get())
			else:
				fSup.get().rightChild.set(supLeftMax.get().leftChild.get())


def removeKey(bst, key):
	remove_i(bst.root, key)
